### ISSUE TITLE
Don't send push notifications to users who are clearly active on a device

### ISSUE THREAD

**Issue by dbkr** [vector-im/element-meta#360]

**Comment by rwjack** [comment]
Any updates on this?
I'm confident it is possible to resolve this if enough energy is directed towards the issue.
We're literally being chased by 7 year old bug, as if it were a ghost!

**Comment by kellyclowers** [comment]
Yeah this is a big problem for me, extremely annoying to a sound sensitive person like myself to be chatting on the computer with someone and their responses are making my phone beep at me.

(also I think there should be a short delay + roll up period, so if someone types many short messages in quick succession, you only get one sound for notification instead of 5 or 6 or whatever... but that's another issue)

**Comment by Cyberes** [comment]
We're at $230 USD.

https://app.bountysource.com/issues/28824980-don-t-send-push-notifications-to-users-who-are-clearly-active-on-a-device

![image](https://user-images.githubusercontent.com/64224601/223525445-cae6e472-f376-418c-aef4-49daa9389590.png)

**Comment by rwjack** [comment]
![image](https://user-images.githubusercontent.com/59068073/227793995-5ab7f5f6-4ddc-4a85-a74b-2485f985191e.png)

**Comment by ghost** [comment]
Wow.. the issue is nearly 8 Years old and still unresolved...

And it still makes the user experience awful..

**Comment by dtheodor** [comment]
Why is this S-Minor? You are a messenger, getting notifications right is an important requirement don't you think? Teams and Slack do this right.

**Comment by marinkol** [comment]
After convincing my upper management,  to try out matrix, to replace our current messenger,   I got most of the things working as people expected.    (well after convincing everyone that gif's were not business critical feature )

But this here is a deal breaker. Given everyone has at least phone and laptop, where they want to access such messages, constant notifications drives everyone insane.   And I haven't even integrated ticketing system into matrix jet.

Will have to put migration on hold until this is fixed, or better client for matrix comes along.

**Comment by Cyberes** [comment]
@marinkol Sounds like they could have possibly lost your business. You should send them a message to the commercial part of Element to let them know how big a deal this is: https://try.element.io/contact-sales

**Comment by rwjack** [comment]
Someone very based just added $500 to the bounty.

https://app.bountysource.com/issues/28824980-don-t-send-push-notifications-to-users-who-are-clearly-active-on-a-device

![image](https://user-images.githubusercontent.com/59068073/237047873-b39ef31c-d616-4057-bcb1-b54b2fc957a1.png)

**Comment by pszypowicz** [comment]
It looks like there is related issue too: https://github.com/vector-im/element-meta/issues/1096

edit: this is the only issue (which I keep as "subscribed") I care about in matrix. I dropped my conversation from it and moved to whats up (which I hate for multiple reason - but they at least handle essentials right: notification system). I can't recommend matrix/element any more, unless that gets solved.

**Comment by K4LCIFER** [comment]
I'm not sure why this has the "Minor" severity tag. The issue of not detecting that another app is in use makes notifications borderline unuseable to the point where I have to keep them permanently turned off -- I can't stand the constant notification alerts on my phone when I'm talking through the desktop app.

**Comment by K4LCIFER** [comment]
> @rwjack 
> Someone very based just added $500 to the bounty.
> 
> https://app.bountysource.com/issues/28824980-don-t-send-push-notifications-to-users-who-are-clearly-active-on-a-device
> 
> ![image](https://user-images.githubusercontent.com/59068073/237047873-b39ef31c-d616-4057-bcb1-b54b2fc957a1.png)

That's rather unfortunate.  As [you can see](https://github.com/bountysource/core/issues) in the current list of issues for BountySource, they have broken all trust with the community, and are being accused of stealing everyone's funds, as shown below:
![image](https://github.com/vector-im/element-meta/assets/7829231/c57d18c4-1313-4ea3-b94b-440b2a7d6cc9)

**Comment by FlitPix** [comment]
I agree that this should be prioritized more - why is this in S-Minor? Marking things as read whenever I move to Element Android is very annoying.

**Comment by Cyberes** [comment]
My understanding is that this is linked to the whole presence issue. I assume that that has to be solved before they can touch this notification issue. Both should be marked as critical importance.

**Comment by GGORG0** [comment]
Hi, are there any updates on this? Has there been any work done?
It's an 8 year old issue at this point and I'm surprised such a "simple" (idk how simple it actually is) thing hasn't been done yet...

**Comment by jan-Sanku** [comment]
This should be S-Critical at this point! I can only speak for myself, but I'd consider this _essential_ for a messaging application. Element is great, otherwise, but over 8.5 years for this is pretty damn bad!

And surely it can't be _that_ hard; Discord, Steam, and pretty much everything else already handle this behavior just fine.

A simple method of detecting whether a user is active on a device is to check for mouse and/or keyboard input globally. You can try this with the aforementioned Steam and Discord: Have your status set to "Online", don't touch your computer for 10 minutes, now your status will be "Idle". Touch your mouse and they both jump back to online. (Discord has a separate (adjustable!) timer for how long it should hold notifications, but that's beside the point)

If all web/desktop clients have not detected any input for n minutes, start sending notifications to mobile devices. If any web/desktop client is active, do not. Simple as that.

**Comment by t3chguy** [comment]
> If all web/desktop clients have not detected any input for n minutes, start sending notifications to mobile devices. If any web/desktop client is active, do not. Simple as that.

The sending of notifications to mobile devices is done by the Matrix server and nothing in the Matrix spec says to behave as you describe, so this is going to be something which someone needs to write an MSC for and then tweak the behaviour in whichever Matrix server implementation you are using.

**Comment by rwjack** [comment]
For anyone brave enough to write that MSC, here's some food for thought:

Since this is a "presence" thing, what happens if servers have presence disabled? 

For the changes that will fix this issue to work, presence should be "server based", which could control the logic of the push notification delivery system, and if servers decide to disable presence, that shouldn't break the push notifications, it's just that the presence status won't be delivered to end clients.

Then again, with "server based" - all the time presence, we're running into some potential privacy issues, such as easier admin snooping of who is online, even if "end user presence" is technically disabled.

**Comment by laurensk** [comment]
Wouldn't it be relatively easy possible to create a `should_notify_event_for_user` callback so we could have a module that listens in before push notifications are sent to a device? We could then use presence to check if a user is online on Element, and if so, discard the push notification.

**Comment by t3chguy** [comment]
Most large Synapse deployments have Presence disabled so such an approach would not function

**Comment by laurensk** [comment]
Having a custom module running that does this is not targeted for large installs. I think many people are thinking of deploying a homeserver internally for their company, which will most definitely have Presence enabled to more closely resemble Slack, Teams, etc...

In my thinking such a callback could be relatively easily implemented, with which we as devs could experiment with solving this problem.

**Comment by AlexanderLuck** [comment]
Very unfortunate that after 11 years this at least from my perspective very basic feature is not working. 
I was hoping to move to Matrix from Slack but within 30 minutes of trying it this immediately was extremely annoying with constant notifcations popping up on my mobile while chatting on Desktop that we've decided to get rid of Matrix again. 
This should be a very high priority and for a first time user this makes it feel like noone is actually using the tool in real life.

**Comment by amstan** [comment]
[Here is my PR](https://github.com/element-hq/element-web/pull/31986) that addresses this for just element-web, a similar patch is needed for the mobile version.

**Comment by MurzNN** [comment]
I think that this issue will be easier to fix on the server side, not on the client side, because the server has the information about all connected devices and read markers, so can easily detect that one of the clients already read the message, or guess if the client "active" and ready to read the new message right now, and just postpone push notifications to other devices for several seconds to wait for the read status from the "active" client.

What do you think about this kind of implementation? Sounds like it is better to discuss this implementation as [MSC](https://github.com/matrix-org/matrix-spec-proposals)?

P.S. By the way, here https://github.com/matrix-org/matrix-spec-proposals/issues/1119 is my MSC1119 that was closed in favor of this issue 😕

**Comment by MurzNN** [comment]
And let's try to help move on with this feature by sharing the details on which other messengers have this feature already implemented, and some details of their technical (or, at least, logical) implementation, if possible.

**Comment by amstan** [comment]
> but realistically I think we need an approach that considers all clients,

> I think that this feature will be easier to implement on the server side

Guys... whatever it takes. But from my point of view I'm here looking at a bug that hasn't been touched in years, I've already been using [my fairly simple fix](https://github.com/element-hq/element-web/pull/31986) for a few months, meanwhile I think the server side arguments (despite maybe being doable) seem familiar to what has been discussed a few times in the other thread years back but no success.

I'm still confused how this is solvable server side with no user presence information.

**Comment by amstan** [comment]
> if the client "active" and ready to read the new message right now, and just postpone push notifications to other devices for several seconds to wait for the read status from the "active" client.

So the server needs to have a list of clients and their active state, until then does it just randomize which client it sends to first?

**Comment by MurzNN** [comment]
> So the server needs to have a list of clients and their active state, until then does it just randomize which client it sends to first?

The server knows the last sync timestamp for each client, and can just send the push notification to the recent one (or to the group of devices that were active in the last 300 secs, eg), and postpone sending to the next recent devices for N seconds, waiting while this client marks the message as read. If it doesn't mark, continue sending the push notification to the next device, and so on, until all devices are notified.

**Comment by nirgilboa** [comment]
hey! Is this on the roadmap to be implemented? and btw is there a public roadmap available somewhere? 🙏🏻

**Comment by kreesty** [comment]
Please can anything be done about this? It's been nearly a decade!

**Comment by alyx161** [comment]
Sad to see this is still an issue over a decade later
